### PR TITLE
Name change settings

### PR DESCRIPTION
**Pull request by Aliblas**
I just made a whole big interface for editing your display name.
Not sure if we'll get around to actually displaying author names by their recipes, but I also wrote a helper method for fetching them with data store for a given authorID. It is in utils/Users.java.
It might be more efficient to get author names with some datastore indexing magic, rather than fetching the whole User entity, but I don't know indexing magic.
edit: author name appears in Recipe objects and will fetch to front end.

CL:
Default display name is anonymous to respect privacy.
settings page shows current display name, allows submission of new name.
settings servlet checks and allows specific characters and patterns.
Front end gives feedback based on name submission.
settings link added to user dropdown.
image

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/UserSettingsServlet.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        json.addProperty("message", "Invalid name input");
+      }
+      json.addProperty("success", false);
+    }
+
+    response.setContentType("application/json");
+    response.getWriter().print(json.toString());
+  }
+
+  boolean hasRepeatingSpaces(String word) {
```
Comment (isamitre) [inline src/main/java/com/google/sps/servlets/UserSettingsServlet.java]: can't you just do indexOf("  ")? if it returns -1 then there aren't any repeating spaces

**Review by isamitre — DISMISSED** [review]
LGTM except for my comment about indexOf
a picture would be nice too

File: src/main/java/com/google/sps/servlets/UserSettingsServlet.java (hunk 1 above)
Comment (Aliblas) [reply src/main/java/com/google/sps/servlets/UserSettingsServlet.java]: I don't think that works just from chedcking indexOf documentation. The function seams to work so I think I'll stick with it.

**Review by isamitre — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/UserSettingsServlet.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String idToken = (String) request.getParameter("idToken");
+    User user = new User(idToken);
+    JsonObject json = new JsonObject();
+
+    String nameInput = (String) request.getParameter("nameInput");
+    nameInput.trim();
```
Comment (oturner24) [inline src/main/java/com/google/sps/servlets/UserSettingsServlet.java]: trim() returns a modified String, it doesn't modify it in place. I think you could combine this all in one if you wanted:
`String nameInput = ((String) request.getParameter("nameInput")).trim();`

File: src/main/java/com/google/sps/servlets/UserSettingsServlet.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+// So far this servlet allows fetching and uploading a display name.
+@WebServlet("/settings")
+public class UserSettingsServlet extends HttpServlet {
+  // This regex checks that user names are alphanumeric and (5 - 25) characters.
+  static final String namePattern = "^[\\w_ -]{5,25}";
```
Comment (oturner24) [inline src/main/java/com/google/sps/servlets/UserSettingsServlet.java]: nit: can you use all caps for constant names?
https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names

File: src/main/java/com/google/sps/servlets/UserSettingsServlet.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    } else {
+      if (!nameInput.matches(namePattern)) {
+        json.addProperty("message", "Invalid character in requested name");
+      } else if (hasRepeatingSpaces(nameInput)) {
+        json.addProperty("message", "Consecutive spaces not allowed.");
+      } else {
+        json.addProperty("message", "Invalid name input");
+      }
+      json.addProperty("success", false);
+    }
```
Comment (oturner24) [inline src/main/java/com/google/sps/servlets/UserSettingsServlet.java]: Up to you, but you could make this more efficient by doing the checks just once:
```
if (!nameInput.matches(namePattern) {
   // set the properties you need
} else if (hasRepeatingSpaces(nameInput) {
   // set the properties you need
} else {
   // here you know it's valid, so you can go ahead and set the name
}

File: src/main/java/com/google/sps/servlets/UserSettingsServlet.java (hunk 1 above)
Comment (oturner24) [inline src/main/java/com/google/sps/servlets/UserSettingsServlet.java]: Is this checking if there are 2 or more spaces in a row in the string? Could you use `word.contains("  ")` instead? (where there are 2 spaces between the quotation marks)

**Review by oturner24 — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/UserSettingsServlet.java (hunk 1 above)
Comment (Aliblas) [reply src/main/java/com/google/sps/servlets/UserSettingsServlet.java]: Oh that works, appalling that I didn't think of that.

**Pull request merged**